### PR TITLE
feat!: allow firebase/php-jwt v6

### PR DESCRIPTION
**Pull request by bshaffer**
see CVE-2021-46743 and https://github.com/advisories/GHSA-8xf4-w7qw-pjjw

### CODE REVIEW

**Comment by ahthomsen** [comment]
Bump - would be great to get merged in to enable meeting security SLAs!

**Comment by bshaffer** [comment]
@ahthomsen for the record, firebase PHP-JWT offers a fix for this in [v5.5](https://github.com/firebase/php-jwt/releases/tag/v5.5.0), which this library supports.

File: src/OAuth2.php [hunk 1]
```diff
@@ -1376,23 +1378,60 @@ private function coerceUri($uri)
 
     /**
      * @param string $idToken
-     * @param string|array|null $publicKey
-     * @param array $allowedAlgs
+     * @param Key|Key[]|string|array|null $publicKey
+     * @param string|array $allowedAlg
```
[the remark this answers is not in the document]
Comment (vishwarajanand) [reply src/OAuth2.php:1382]:
```suggestion
     * @param string|array $allowedAlgos
```

File: src/OAuth2.php [hunk 2]
```diff
@@ -1376,23 +1378,60 @@ private function coerceUri($uri)
...
      * @return object
      */
     private function jwtDecode($idToken, $publicKey, $allowedAlgs)
     {
-        return JWT::decode($idToken, $publicKey, $allowedAlgs);
+        $keys = $this->getFirebaseJwtKeys($publicKey, $allowedAlgs);
+
+        // Default exception if none are caught
```
Comment (vishwarajanand) [inline src/OAuth2.php:1389]:
```suggestion
        // Default exception if all are caught
```

File: src/OAuth2.php [hunk 3]
```diff
@@ -1376,23 +1378,60 @@ private function coerceUri($uri)
...
-            $signingKey,
-            $signingAlgorithm,
-            $signingKeyId
-        );
+        // If $allowedAlgs is empty, assume $publicKey is Key or Key[].
+        if (empty($allowedAlgs)) {
+            return (array) $publicKey;
+        }
+
+        if (is_string($allowedAlgs)) {
```
Comment (vishwarajanand) [inline src/OAuth2.php:1411]: I think `$allowedAlgs` check could instead check against all the supported algos mentioned [here](https://github.com/firebase/php-jwt/blob/main/src/JWT.php#L56-L66).

```
    public static $supported_algs = [
        'ES384' => ['openssl', 'SHA384'],
        'ES256' => ['openssl', 'SHA256'],
        'HS256' => ['hash_hmac', 'SHA256'],
        'HS384' => ['hash_hmac', 'SHA384'],
        'HS512' => ['hash_hmac', 'SHA512'],
        'RS256' => ['openssl', 'SHA256'],
        'RS384' => ['openssl', 'SHA384'],
        'RS512' => ['openssl', 'SHA512'],
        'EdDSA' => ['sodium_crypto', 'EdDSA'],
    ];
```

**Review by vishwarajanand — CHANGES_REQUESTED** [review]

File: src/OAuth2.php (hunk 2 above)
Comment (bshaffer) [reply src/OAuth2.php:1389]: Actually I think what i have here is better. The reason is that the API exception being thrown is the default to preserve backwards compatibility. The validation doesn't happen here, but we pretend like it does for BC.

File: src/OAuth2.php [hunk 4]
```diff
@@ -1376,23 +1378,60 @@ private function coerceUri($uri)
...
+     * @param string|array $allowedAlg
      * @return object
      */
     private function jwtDecode($idToken, $publicKey, $allowedAlgs)
     {
-        return JWT::decode($idToken, $publicKey, $allowedAlgs);
+        $keys = $this->getFirebaseJwtKeys($publicKey, $allowedAlgs);
+
+        // Default exception if none are caught
+        $e = new \InvalidArgumentException('Key may not be empty');
```
[the remark this answers is not in the document]
Comment (bshaffer) [reply src/OAuth2.php:1390]: It's the same here .. this exception class and message were taken from firebase/php-jwt. I agree the exception and message don't make sense here, but they are here for backwards compatibility. I will add a comment to make this more clear.

File: src/OAuth2.php (hunk 3 above)
Comment (bshaffer) [reply src/OAuth2.php:1478]: This check is already done in `firebase/php-jwt`, so there is no need to do that check here... we only want to verify type-safety and make sure we handle the types correctly before calling `Firebase\JWT\JWT::decode`.

File: src/OAuth2.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (bshaffer) [reply src/OAuth2.php:1457]: Comment added!

**Comment by bshaffer** [comment]
@dwsupplee @vishwarajanand comments addressed! PTAL!

**Review by vishwarajanand — APPROVED** [review]
LGTM but please have @dwsupplee approve as well.

**Review by dwsupplee — APPROVED** [review]
Looks great! Thanks for the tests 👍

**Comment by bshaffer** [comment]
This has been released in [v1.21.0](https://github.com/googleapis/google-auth-library-php/releases/tag/v1.21.0)

**Pull request merged**